File: BinarySearch.java
/*
Binary search algorithm implemenation
Space complexity = O(1)
Time complexity = O(logn)
*/
public class BinarySearch{
	
	private static int search(int[] arr, int num, int l, int r){
		int mid = l+(r-l)/2;
		if (r>=l)
		{
			if (arr[mid]==num)
			{
				return mid+1;
			}
			else if (arr[mid]<num)
			{
				return search(arr, num, mid+1, r);
			}
			else if (arr[mid]>num)
			{
				return search(arr, num, l, mid-1);
			}
		}
		return -1;
	}

	public static void main(String[] args){
		// Prerequisite for binary search algorithm is that the array must be sorted.
		int arr[] = {4, 28, 31, 45, 47, 50, 59, 78, 80, 93};

		System.out.printf("Num %d is at position %d in the given array.\n",47,search(arr, 47, 0, arr.length));
		System.out.printf("Num %d is at position %d in the given array.",5,search(arr, 5, 0, arr.length));
	}
}

File: BinarySearch.py
'''
Implementation of binary search algorithm in Python
Space complexity = O(1)
Time complexity = O(logn)
Prerequisites - the array passed to the binarysearch method must be sorted
'''
def binarysearch(arr,l,r,x):
    mid=l+(r-l)//2
    if (r>=l):
        if (arr[mid]==x):
            return mid
        elif (arr[mid]>x):
            return binarysearch(arr,l,mid-1,x)
        else:
            return binarysearch(arr,mid+1,r,x)
    return -1

arr1=[1,10,23,42,56,78,90,106]
print("42 is at position ",binarysearch(arr1,0,len(arr1),42)," in the given array")
print("4 is at position ",binarysearch(arr1,0,len(arr1),4)," in the given array")
